src/language-js/print/function.js

```diff
@@ -249,160 +248,243 @@ function printArrowFunctionSignature(path, options, print, args) {
+  // we should align the expression's closing } with the line with the opening {.
+  const shouldAddSoftLine =
+    (args?.expandLastArg || parent.type === "JSXExpressionContainer") &&
+    !hasComment(node);
+
+  let decoratedBodyDoc;
+  if (shouldAlwaysAddParens) {
+    decoratedBodyDoc = group(["(", indent([softline, bodyDoc]), softline, ")"]);
+  } else if (shouldAddParensIfNotBreak && shouldPutBodyOnSameLine) {
+    decoratedBodyDoc = group([
+      ifBreak("", "("),
+      indent([softline, bodyDoc]),
+      ifBreak("", ")"),
+      shouldAddSoftLine
+        ? [ifBreak(printTrailingComma ? "," : ""), softline]
+        : "",
+    ]);
+  } else {
+    decoratedBodyDoc = bodyDoc;
+  }
+  return shouldPutBodyOnSameLine
+    ? [" ", decoratedBodyDoc, bodyComments]
+    : [
+        indent([line, decoratedBodyDoc, bodyComments]),
+        shouldAddSoftLine
+          ? [ifBreak(printTrailingComma ? "," : ""), softline]
+          : "",
+      ];
 }
 
 function printArrowFunction(path, options, print, args) {
-  let { node } = path;
+  const { node, parent, key } = path;
   /** @type {Doc[]} */
   const signatures = [];
-  const body = [];
+  /** @type {Doc} */
+  let bodyDoc;
+  /** @type {*[]} */
+  const bodyComments = [];
   let chainShouldBreak = false;
+  let tailNode = node;
 
   (function rec() {
+    const { node: currentNode } = path;
     const doc = printArrowFunctionSignature(path, options, print, args);
     if (signatures.length === 0) {
       signatures.push(doc);
     } else {
       const { leading, trailing } = printCommentsSeparately(path, options);
       signatures.push([leading, doc]);
-      body.unshift(trailing);
+      bodyComments.unshift(trailing);
     }
 
     chainShouldBreak =
       chainShouldBreak ||
       // Always break the chain if:
-      (node.returnType && getFunctionParameters(node).length > 0) ||
-      node.typeParameters ||
-      getFunctionParameters(node).some((param) => param.type !== "Identifier");
+      (currentNode.returnType &&
+        getFunctionParameters(currentNode).length > 0) ||
+      currentNode.typeParameters ||
+      getFunctionParameters(currentNode).some(
+        (param) => param.type !== "Identifier"
+      );
 
-    if (node.body.type !== "ArrowFunctionExpression" || args?.expandLastArg) {
-      body.unshift(print("body", args));
+    if (
+      currentNode.body.type !== "ArrowFunctionExpression" ||
+      args?.expandLastArg
+    ) {
+      bodyDoc = print("body", args);
     } else {
-      node = node.body;
+      tailNode = currentNode.body;
       path.call(rec, "body");
     }
   })();
 
-  if (signatures.length > 1) {
-    return printArrowChain(
-      path,
-      args,
-      signatures,
-      chainShouldBreak,
+  const { body } = tailNode;
+
+  // In order to avoid confusion between
+  // a => a ? a : a
+  // a <= a ? a : a
+  const shouldAddParensIfNotBreak =
+    body.type === "ConditionalExpression" &&
+    !startsWithNoLookaheadToken(
```

Comment:
This is an expensive call, can we avoid it if possible?

Reply:
I'm not sure if I understand your comment correctly.
Anyway, just eliminating this call looks reasonable.
It looks to improve not only performance but simplicity of code and readability of result.
```js
// Input
() => ({} ? b : c);
a => ({}::b()``[''].c++ && 0 ? 0 : 0);

// Stable
() => ({} ? b : c);
(a) => ({}::b()``[""].c++ && 0 ? 0 : 0);

// Next (difference against stable probably comes from #14599)
() => ({}) ? b : c;
(a) => ({})::b()``[""].c++ && 0 ? 0 : 0;

// Before eliminating it in this branch
() => ({}) ? b : c;
(a) => ({})::b()``[""].c++ && 0 ? 0 : 0;

// If we eliminate the call
() => (({}) ? 1 : 2);
(a) => (({})::b()\`\`[""].c++ && 0 ? 0 : 0);
```

Reply:
Or, I wonder if we can add parens for conditional expression even if it breaks like sequence expression.
Then, the logic will get more simpler and we might reduce chance of call of `startsWithNoLookaheadToken()` adding `arrowFunctionBody.type !== "ConditionalExpression"`.

Reply:
Though it might be better, I also think that it looks out of scope to add parens for conditional expression even if it breaks.
It will change output for so many practical codes.

Reply:
Sorry for didn't express good. I didn't mean to change logic, but don't all it when we don't need it. I'll take a look.

Reply:
I'm sorry but I don't understand it yet... 🙇 
At the previous revision, `startsWithNoLookaheadToken()` weren't called if `body.type === "ConditionalExpression"` is falsy thanks to short-circuit evaluation, and, I think, avoiding unnecessary call was already achieved.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND#short-circuit_evaluation

Reply:
Now if `body.type === 'SequenceExpression'` or `shouldPutBodyOnSameLine`, we don't need call it [`8ef4b8d` (#14633)](https://github.com/prettier/prettier/pull/14633/commits/8ef4b8d49aedd938a54587a291dcdfe638987e5e).

Make sense?

Reply:
You might miss read my comment, by comment is about "SequenceExpression", not "ConditionalExpression" 😄

Reply:
Thank you so much for friendly description.
I missed `shouldPutBodyOnSameLine`.
I meant, if `body.type === 'SequenceExpression'`, `body.type === "ConditionalExpression"` should be false and the function isn't called.
I'm still on the way to understanding it, but it looks to make sense.